nvflare/fuel/hci/client/api.py

```diff
@@ -421,22 +333,151 @@ def __init__(
+            client_name=self.user_name,
+            client_type=ClientType.ADMIN,
+            expected_sp_identity=self.server_identity,
+            secure_mode=True,  # always True to authenticate the cell endpoint!
+            root_cert_file=self.ca_cert,
+            private_key_file=self.client_key,
+            cert_file=self.client_cert,
+            msg_timeout=2.0,
+            retry_interval=1.0,
+            timeout=timeout,
+        )
+
+        abort_signal = Signal()
+        shared_fl_ctx = FLContext()
+        shared_fl_ctx.set_public_props({ReservedKey.IDENTITY_NAME: self.user_name})
+        token, token_signature, ssid, token_verifier = authenticator.authenticate(
+            shared_fl_ctx=shared_fl_ctx,
+            abort_signal=abort_signal,
+        )
+
+        if not isinstance(token_verifier, TokenVerifier):
+            raise RuntimeError(f"expect token_verifier to be TokenVerifier but got {type(token_verifier)}")
+
+        set_add_auth_headers_filters(self.cell, self.user_name, token, token_signature, ssid)
+
+        self.cell.core_cell.add_incoming_filter(
+            channel="*",
+            topic="*",
+            cb=validate_auth_headers,
+            token_verifier=token_verifier,
+            logger=self.logger,
+        )
+        self.debug(f"Successfully authenticated to {self.server_identity}: {token=} {ssid=}")
+
+        self.aux_runner = AuxRunner(self)
+        self.object_streamer = ObjectStreamer(self.aux_runner)
+
+        self.cell.register_request_cb(
+            channel=CellChannel.AUX_COMMUNICATION,
+            topic="*",
+            cb=self._handle_aux_message,
+        )
+
+    def _handle_aux_message(self, request: CellMessage) -> CellMessage:
+        assert isinstance(request, CellMessage), "request must be CellMessage but got {}".format(type(request))
+        data = request.payload
+
+        topic = request.get_header(MessageHeaderKey.TOPIC)
+        with self.new_context() as fl_ctx:
+            reply = self.aux_runner.dispatch(topic=topic, request=data, fl_ctx=fl_ctx)
+
+            if reply is not None:
+                return_message = CellMessage({}, reply)
+                return_message.set_header(MessageHeaderKey.RETURN_CODE, CellReturnCode.OK)
+            else:
+                return_message = CellMessage({}, None)
+            return return_message
+
+    def download_file(self, source_fqcn: str, ref_id: str, file_name: str):
+        err, file_path = downloader.download_file(
+            cell=self.cell,
+            ref_id=ref_id,
+            from_fqcn=source_fqcn,
+            per_request_timeout=self.file_download_progress_timeout,
+        )
+        if err:
+            print(f"failed to receive file {file_name}: {err}")
+            return None
+
+        file_stats = os.stat(file_path)
+        num_bytes_received = file_stats.st_size
+        Path(os.path.dirname(file_name)).mkdir(parents=True, exist_ok=True)
+        shutil.move(file_path, file_name)
```

Comment:
some logic of renaming/removing tempfile/folder is handled here, should we move them into the downloader.download_file?

Reply:
Not really. The lifecycle of the downloaded file should be managed by the caller, not the downloader.